adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/impl/VertxBasedAmqpProtocolAdapter.java

```diff
@@ -1250,23 +1255,37 @@ private void reportSentCommand(final TenantObject tenantObject, final CommandCon
         }
     }
 
-    private static void setConnectionLossHandler(final ProtonConnection con, final Handler<Void> handler) {
-
-        con.attachments().set("connectionLossHandler", Handler.class, handler);
+    private static void setConnectionLossHandler(final ProtonConnection con, final String key, final Function<Span, Future<Void>> handler) {
+        @SuppressWarnings("unchecked")
+        final Map<String, Function<Span, Future<Void>>> handlers = Optional
+                .ofNullable(con.attachments().get(KEY_CONNECTION_LOSS_HANDLERS, Map.class))
+                .orElse(new HashMap<>());
+        handlers.put(key, handler);
+        con.attachments().set(KEY_CONNECTION_LOSS_HANDLERS, Map.class, handlers);
     }
 
-    @SuppressWarnings("unchecked")
-    private static Handler<Void> getConnectionLossHandler(final ProtonConnection con) {
+    private static Collection<Function<Span, Future<Void>>> getConnectionLossHandlers(final ProtonConnection con) {
+        @SuppressWarnings("unchecked")
+        final Map<String, Function<Span, Future<Void>>> handlers = con.attachments().get(KEY_CONNECTION_LOSS_HANDLERS, Map.class);
+        return handlers != null ? handlers.values() : Collections.emptyList();
+    }
 
-        return con.attachments().get("connectionLossHandler", Handler.class);
+    private static boolean removeConnectionLossHandler(final ProtonConnection con, final String key) {
+        @SuppressWarnings("unchecked")
+        final Map<String, Function<Span, Future<Void>>> handlers = con.attachments().get(KEY_CONNECTION_LOSS_HANDLERS, Map.class);
+        return handlers != null && handlers.remove(key) != null;
     }
 
     private static Device getAuthenticatedDevice(final ProtonConnection con) {
-        return Optional.ofNullable(con.attachments())
-                .map(attachments -> attachments.get(AmqpAdapterConstants.KEY_CLIENT_DEVICE, Device.class))
+        return Optional.ofNullable(con.attachments().get(AmqpAdapterConstants.KEY_CLIENT_DEVICE, Device.class))
```

Comment:
does *ProtonConnection.attachments()* always return a non-null value?

Reply:
With the current Proton implementation classes, yes. And besides, it would be inconsistent having the null check here, but not in the many other places where the attachments are used.